### PR TITLE
drivers/hd44780: fix potential hardfault during initialization

### PR DESCRIPTION
**Pull request by aabadie**
### Contribution description

This PR fixes a hardfault when running the `tests/driver_hd44780` application on samr21-xpro, when built with LLVM, as reported in https://github.com/RIOT-OS/RIOT/pull/12487#issuecomment-547850143.

In fact, the test application also crashes when run on arduino-zero/samr30-xpro if built GCC or LLVM.

Since the data pins array param is stored on flash memory and the driver is accessing it in loops for initializing/setting/clearing the pins. I thought that there was something with this that could cause a crash so this PR is just adding a extra array in RAM for and use it in the driver instead of the param array. It seems to work, at least it doesn't crash anymore.

Surprisingly, there's no RAM increase, just 24 extra bytes on ROM.

There might be something else, related to SAM0 platforms since it works well on the other platforms I tested (STM32, nRF), but I have no idea what it could be.

### Testing procedure

- Run
  ```
  $ make BOARD=samr21-xpro -C tests/driver_hd44780 flash test
  $ TOOLCHAIN=llvm make BOARD=samr21-xpro -C tests/driver_hd44780 flash test
  ```
  on master and with this PR the first one works but the second one only works with this PR.
  Changing BOARD for arduino-zero or samr30-xpro none of the 2 commands works on master but both are fixed by this PR.

### Issues/PRs references

Fixes the issue reported in  https://github.com/RIOT-OS/RIOT/pull/12487#issuecomment-547850143

### CODE REVIEW

File: drivers/hd44780/hd44780.c [hunk 1]
```diff
@@ -110,19 +110,17 @@ int hd44780_init(hd44780_t *dev, const hd44780_params_t *params)
...
         return -1;
     }
-    uint8_t count_pins = 0;
-    /* check which pins are used */
+    /* Copy configured pins in device descriptor */
     for (unsigned i = 0; i < HD44780_MAX_PINS; ++i) {
-        if (dev->p.data[i] != GPIO_UNDEF) {
-            ++count_pins;
-        }
+        dev->data[i] = dev->p.data[i];
```
Comment (MrKevinWeiss) [inline drivers/hd44780/hd44780.c]: any reason no `memcpy(dev->data, dev.p.data, ARRAY_SIZE((dev->data));`  (I don't know if that is the proper array size macro but you get the idea)?

Comment (aabadie) [reply drivers/hd44780/hd44780.c]: I tried memcpy and noticed a bigger code size on ROM, that's why I kept the for loop version.

Comment (MrKevinWeiss) [reply drivers/hd44780/hd44780.c]: Fair enough, lets me run it through the test.

**Comment by aabadie** [comment]
> it works in GCC and I don't see anything obviously wrong with it. After all gpio_t is just a uint32_t, it shouldn't matter where it resides.

While working on it, I discovered that saml21 based boards are broken with GCC, same for Arduino-Zero (samd21). So there's something wrong with sam0 boards in general, it's just that samr21 is less affected I would say. On the other architectures I tested (STM32, nRF51) the test is working.

> How did you set up the llvm tookchain?

Just installed llvm locally but you can also use this toolchain from within docker.

**Comment by benpicco** [comment]
I can't even build with `TOOLCHAIN=llvm` when I select `BOARD=same54-xpro`:
```
/home/benpicco/dev/RIOT/cpu/cortexm_common/thread_arch.c:309:6: error: instruction requires: VFP2
    "vstmdbeq r0!, {s16-s31}          \n" /* save FPU registers if FPU is used */
     ^
<inline asm>:5:1: note: instantiated into assembly here
vstmdbeq r0!, {s16-s31}          
^
/home/benpicco/dev/RIOT/cpu/cortexm_common/thread_arch.c:311:6: error: incorrect condition in IT block; got 'al', but expected 'eq'
    "stmdb  r0!,{r4-r11}              \n" /* save regs */
     ^
<inline asm>:6:6: note: instantiated into assembly here
stmdb  r0!,{r4-r11}              
     ^
/home/benpicco/dev/RIOT/cpu/cortexm_common/thread_arch.c:360:6: error: instruction requires: VFP2
    "vldmiaeq r1!, {s16-s31}          \n" /* load FPU registers if saved */
     ^
<inline asm>:12:1: note: instantiated into assembly here
vldmiaeq r1!, {s16-s31}          
^
/home/benpicco/dev/RIOT/cpu/cortexm_common/thread_arch.c:362:6: error: incorrect condition in IT block; got 'al', but expected 'eq'
    "msr    psp, r1                   \n" /* restore user mode SP to PSP reg */
     ^
<inline asm>:13:4: note: instantiated into assembly here
msr    psp, r1                   
   ^
4 errors generated.
```

When I use `BOARD=samr21-xpro` the build succeeds oO

**Comment by aabadie** [comment]
I had the same issue with nrf52. Maybe the toolchain just doesn't support M4 ?

**Comment by MrKevinWeiss** [comment]
A few boards have failed though I don't know if we should expect failures for this test since the boards don't have the driver connected... Maybe some GPIO pins are not valid in the boards or something:

<details><summary>ek-lm4f120xl</summary>

```
make RIOT_CI_BUILD=1 CC_NOCOLOR=1 --no-print-directory -C ./tests/driver_hd44780 test
/srv/ilab-builds/workspace/jobs/experimental-pull-request-tests/dist/tools/pyterm/pyterm -p "/dev/riot/tty-ek-lm4f120xl" -b "115200" --noprefix --no-repeat-command-on-empty-line
Twisted not available, please install it if you want to use pyterm's JSON capabilities
Connect to serial port /dev/riot/tty-ek-lm4f120xl
Welcome to pyterm!
Type '/exit' to exit.
main(): This is RIOT! (Version: buildtest)
[START]

Context before hardfault:
   r0: 0x00000000
   r1: 0x00000000
   r2: 0x40004000
   r3: 0x00004000
  r12: 0x00000000
   lr: 0x000014fb
   pc: 0x000014fa
  psr: 0x41000000

FSR/FAR:
 CFSR: 0x00000400
 HFSR: 0x40000000
 DFSR: 0x00000008
 AFSR: 0x00000000
Misc
EXC_RET: 0xfffffffd
Attempting to reconstruct state for debugging...
In GDB:
  set $pc=0x14fa
  frame 0
  bt

ISR stack overflowed by at least 16 bytes.
Timeout in expect script at "child.expect_exact("[SUCCESS]")" (tests/driver_hd44780/tests/01-run.py:16)

/srv/ilab-builds/workspace/jobs/experimental-pull-request-tests/tests/driver_hd44780/../../Makefile.include:710: recipe for target 'test' failed
make: *** [test] Error 1

Return value: 2
```
</details>

<details><summary>stm32f3discovery - I think just a uart problem though</summary>

```
make RIOT_CI_BUILD=1 CC_NOCOLOR=1 --no-print-directory -C ./tests/driver_hd44780 test
/srv/ilab-builds/workspace/jobs/experimental-pull-request-tests/dist/tools/pyterm/pyterm -p "/dev/riot/tty-stm32f3discovery" -b "115200" --noprefix --no-repeat-command-on-empty-line
Twisted not available, please install it if you want to use pyterm's JSON capabilities
Connect to serial port /dev/riot/tty-stm32f3discovery
Welcome to pyterm!
Type '/exit' to exit.
\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00main(): This is RIOT! (Version: buildtest)
[START]
Timeout in expect script at "child.expect_exact("[SUCCESS]")" (tests/driver_hd44780/tests/01-run.py:16)

/srv/ilab-builds/workspace/jobs/experimental-pull-request-tests/tests/driver_hd44780/../../Makefile.include:710: recipe for target 'test' failed
make: *** [test] Error 1

Return value: 2
```
</details>

<details><summary>esp32-wroom32</summary>

```
make RIOT_CI_BUILD=1 CC_NOCOLOR=1 --no-print-directory -C ./tests/driver_hd44780 test
ESP32_SDK_DIR should be defined as /path/to/esp-idf directory
ESP32_SDK_DIR is set by default to /opt/esp/esp-idf
/srv/ilab-builds/workspace/jobs/experimental-pull-request-tests/dist/tools/pyterm/pyterm -p "/dev/riot/tty-esp32-wroom-32" -b "115200" --noprefix --no-repeat-command-on-empty-line
Twisted not available, please install it if you want to use pyterm's JSON capabilities
Connect to serial port /dev/riot/tty-esp32-wroom-32
Welcome to pyterm!
Type '/exit' to exit.
ets Jun  8 2016 00:22:57

rst:0x7 (TG0WDT_SYS_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DOUT, clock div:2
load:0x3fff0018,len:4
load:0x3fff001c,len:5632
ho 0 tail 12 room 4
load:0x40078000,len:0
load:0x40078000,len:14868
entry 0x40078628
[0;33mW (30) boot: PRO CPU has been reset by WDT.[0m
[0;33mW (31) boot: WDT reset info: PRO CPU PC=0x4008175f[0m
[0;33mW (31) boot: WDT reset info: APP CPU PC=0x73794bdd[0m
[0;32mI (37) boot: ESP-IDF v3.1-dev-961-ga255622 2nd stage bootloader[0m
[0;32mI (43) boot: compile time 12:38:02[0m
[0;32mI (47) boot: Enabling RNG early entropy source...[0m
[0;32mI (53) boot: SPI Speed      : 40MHz[0m
[0;32mI (57) boot: SPI Mode       : DOUT[0m
[0;32mI (61) boot: SPI Flash Size : 4MB[0m
[0;32mI (65) boot: Partition Table:[0m
[0;32mI (69) boot: ## Label            Usage          Type ST Offset   Length[0m
[0;32mI (76) boot:  0 nvs              WiFi data        01 02 00009000 00006000[0m
[0;32mI (83) boot:  1 phy_init         RF data          01 01 0000f000 00001000[0m
[0;32mI (91) boot:  2 factory          factory app      00 00 00010000 00018ad0[0m
[0;32mI (98) boot: End of partition table[0m
[0;32mI (103) esp_image: segment 0: paddr=0x00010020 vaddr=0x3f400020 size=0x01920 (  6432) map[0m
[0;32mI (114) esp_image: segment 1: paddr=0x00011948 vaddr=0x3ffb0000 size=0x01198 (  4504) load[0m
[0;32mI (122) esp_image: segment 2: paddr=0x00012ae8 vaddr=0x40080000 size=0x00400 (  1024) load[0m
[0;32mI (130) esp_image: segment 3: paddr=0x00012ef0 vaddr=0x40080400 size=0x0467c ( 18044) load[0m
[0;32mI (146) esp_image: segment 4: paddr=0x00017574 vaddr=0x00000000 size=0x08a9c ( 35484) [0m
[0;32mI (160) esp_image: segment 5: paddr=0x00020018 vaddr=0x400d0018 size=0x08a8c ( 35468) map[0m
[0;32mI (176) boot: Loaded app from partition at offset 0x10000[0m
[0;32mI (176) boot: Disabling RNG early entropy source...[0m

Starting ESP32 with ID: 1730aea4d34600

Current clocks in Hz: CPU=80000000 APB=80000000 XTAL=40000000 SLOW=150000
PRO cpu is up (single core mode, only PRO cpu is used)
PRO cpu starts user code
Used clocks in Hz: CPU=80000000 APB=80000000 XTAL
Board configuration:
	UART_DEV(0)	txd=1 rxd=3
	UART_DEV(1)	txd=10 rxd=9
	LED		pins=[ ]
	BUTTONS		pins=[ 0 ]

Starting RIOT kernel on PRO cpu
I (13630) [main_trampoline]: main(): This is RIOT! (Version: buildtest)
[START]
EXCEPTION!! exccause=0 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (EXCEPTION!! exccause=28 (ets Jun  8 2016 00:22:57

rst:0x7 (TG0WDT_SYS_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DOUT, clock div:2
load:0x3fff0018,len:4
load:0x3fff001c,len:5632
ho 0 tail 12 room 4
load:0x40078000,len:0
load:0x40078000,len:14868
entry 0x40078628
[0;33mW (30) boot: PRO CPU has been reset by WDT.[0m
[0;33mW (31) boot: WDT reset info: PRO CPU PC=0x4008175f[0m
[0;33mW (31) boot: WDT reset info: APP CPU PC=0x73794bdd[0m
[0;32mI (37) boot: ESP-IDF v3.1-dev-961-ga255622 2nd stage bootloader[0m
[0;32mI (43) boot: compile time 12:38:02[0m
[0;32mI (47) boot: Enabling RNG early entropy source...[0m
[0;32mI (53) boot: SPI Speed      : 40MHz[0m
[0;32mI (57) boot: SPI Mode       : DOUT[0m
[0;32mI (61) boot: SPI Flash Size : 4MB[0m
[0;32mI (65) boot: Partition Table:[0m
[0;32mI (69) boot: ## Label            Usage          Type ST Offset   Length[0m
[0;32mI (76) boot:  0 nvs              WiFi data        01 02 00009000 00006000[0m
[0;32mI (83) boot:  1 phy_init         RF data          01 01 0000f000 00001000[0m
[0;32mI (91) boot:  2 factory          factory app      00 00 00010000 00018ad0[0m
[0;32mI (98) boot: End of partition table[0m
[0;32mI (103) esp_image: segment 0: paddr=0x00010020 vaddr=0x3f400020 size=0x01920 (  6432) map[0m
[0;32mI (114) esp_image: segment 1: paddr=0x00011948 vaddr=0x3ffb0000 size=0x01198 (  4504) load[0m
[0;32mI (122) esp_image: segment 2: paddr=0x00012ae8 vaddr=0x40080000 size=0x00400 (  1024) load[0m
[0;32mI (130) esp_image: segment 3: paddr=0x00012ef0 vaddr=0x40080400 size=0x0467c ( 18044) load[0m
[0;32mI (146) esp_image: segment 4: paddr=0x00017574 vaddr=0x00000000 size=0x08a9c ( 35484) [0m
[0;32mI (160) esp_image: segment 5: paddr=0x00020018 vaddr=0x400d0018 size=0x08a8c ( 35468) map[0m
[0;32mI (176) boot: Loaded app from partition at offset 0x10000[0m
[0;32mI (176) boot: Disabling RNG early entropy source...[0m

Starting ESP32 with ID: 1730aea4d34600

Current clocks in Hz: CPU=80000000 APB=80000000 XTAL=40000000 SLOW=150000
PRO cpu is up (single core mode, only PRO cpu is used)
PRO cpu starts user code
Used clocks in Hz: CPU=80000000 APB=80000000 XTAL
Board configuration:
	UART_DEV(0)	txd=1 rxd=3
	UART_DEV(1)	txd=10 rxd=9
	LED		pins=[ ]
	BUTTONS		pins=[ 0 ]

Starting RIOT kernel on PRO cpu
I (20318) [main_trampoline]: main(): This is RIOT! (Version: buildtest)
[START]
Timeout in expect script at "child.expect_exact("[SUCCESS]")" (tests/driver_hd44780/tests/01-run.py:16)

/srv/ilab-builds/workspace/jobs/experimental-pull-request-tests/tests/driver_hd44780/../../Makefile.include:710: recipe for target 'test' failed
make: *** [test] Error 1

Return value: 2
```
</details>

**Comment by MrKevinWeiss** [comment]
I can build the same54-xpro with llvm:

<details>

```
TOOLCHAIN=llvm make BOARD=same54-xpro -C tests/driver_hd44780 all
make: Entering directory '/home/kevinweiss/WorkingDirectory/RIOT/tests/driver_hd44780'
Building application "tests_driver_hd44780" for "same54-xpro" with MCU "samd5x".

"make" -C /home/kevinweiss/WorkingDirectory/RIOT/boards/same54-xpro
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/core
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/cpu/samd5x
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/cpu/cortexm_common
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/cpu/cortexm_common/periph
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/cpu/sam0_common
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/cpu/sam0_common/periph
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/cpu/samd5x/periph
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/drivers
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/drivers/hd44780
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/drivers/periph_common
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys/auto_init
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys/div
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys/newlib_syscalls_default
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys/pm_layered
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys/stdio_uart
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys/xtimer
   text	   data	    bss	    dec	    hex	filename
  12084	    120	   2576	  14780	   39bc	/home/kevinweiss/WorkingDirectory/RIOT/tests/driver_hd44780/bin/same54-xpro/tests_driver_hd44780.elf
make: Leaving directory '/home/kevinweiss/WorkingDirectory/RIOT/tests/driver_hd44780'
```

**Comment by aabadie** [comment]
@MrKevinWeiss are the failing boards also failing on master (and eventually the latest release version) ?
Does it work on same54-xpro with LLVM ?

**Comment by MrKevinWeiss** [comment]
Probably good to point that out... Yup failing in master too so maybe we don't worry about it, yet.

**Comment by MrKevinWeiss** [comment]
For the same54 I seem to have some flashing problems, looking into it.

<details>

```
TOOLCHAIN=llvm make BOARD=same54-xpro -C tests/driver_hd44780 flash test
make: Entering directory '/home/kevinweiss/WorkingDirectory/RIOT/tests/driver_hd44780'
Building application "tests_driver_hd44780" for "same54-xpro" with MCU "samd5x".

"make" -C /home/kevinweiss/WorkingDirectory/RIOT/boards/same54-xpro
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/core
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/cpu/samd5x
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/cpu/cortexm_common
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/cpu/cortexm_common/periph
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/cpu/sam0_common
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/cpu/sam0_common/periph
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/cpu/samd5x/periph
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/drivers
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/drivers/hd44780
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/drivers/periph_common
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys/auto_init
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys/div
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys/newlib_syscalls_default
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys/pm_layered
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys/stdio_uart
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys/xtimer
   text	   data	    bss	    dec	    hex	filename
  12084	    120	   2576	  14780	   39bc	/home/kevinweiss/WorkingDirectory/RIOT/tests/driver_hd44780/bin/same54-xpro/tests_driver_hd44780.elf
/home/kevinweiss/WorkingDirectory/RIOT/dist/tools/edbg/edbg  --target atmel_cm4v2 --verbose --file /home/kevinweiss/WorkingDirectory/RIOT/tests/driver_hd44780/bin/same54-xpro/tests_driver_hd44780.bin --verify || /home/kevinweiss/WorkingDirectory/RIOT/dist/tools/edbg/edbg  --target atmel_cm4v2 --verbose --file /home/kevinweiss/WorkingDirectory/RIOT/tests/driver_hd44780/bin/same54-xpro/tests_driver_hd44780.bin --verify --program
Error: unknown target type (atmel_cm4v2); use '-t' option
Error: unknown target type (atmel_cm4v2); use '-t' option
/home/kevinweiss/WorkingDirectory/RIOT/tests/driver_hd44780/../../Makefile.include:650: recipe for target 'flash' failed
make: *** [flash] Error 1
make: Leaving directory '/home/kevinweiss/WorkingDirectory/RIOT/tests/driver_hd44780'

```

**Comment by MrKevinWeiss** [comment]
Bumped the version of edbg and cleared out the binary in `dist/tools/edbg` and reran.  It passes.

<details>

```
make BOARD=same54-xpro -C tests/driver_hd44780 all flash test
make: Entering directory '/home/kevinweiss/WorkingDirectory/RIOT/tests/driver_hd44780'
Building application "tests_driver_hd44780" for "same54-xpro" with MCU "samd5x".

"make" -C /home/kevinweiss/WorkingDirectory/RIOT/boards/same54-xpro
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/core
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/cpu/samd5x
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/cpu/cortexm_common
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/cpu/cortexm_common/periph
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/cpu/sam0_common
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/cpu/sam0_common/periph
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/cpu/samd5x/periph
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/drivers
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/drivers/hd44780
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/drivers/periph_common
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys/auto_init
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys/div
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys/newlib_syscalls_default
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys/pm_layered
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys/stdio_uart
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys/xtimer
   text	   data	    bss	    dec	    hex	filename
  11292	    120	   2576	  13988	   36a4	/home/kevinweiss/WorkingDirectory/RIOT/tests/driver_hd44780/bin/same54-xpro/tests_driver_hd44780.elf
[INFO] edbg binary not found - building it from source now
CC= CFLAGS= make -C /home/kevinweiss/WorkingDirectory/RIOT/dist/tools/edbg
rm -Rf /home/kevinweiss/WorkingDirectory/RIOT/dist/tools/edbg/bin
mkdir -p /home/kevinweiss/WorkingDirectory/RIOT/dist/tools/edbg/bin
/home/kevinweiss/WorkingDirectory/RIOT/dist/tools/git/git-cache clone "https://github.com/ataradov/edbg" "4f5d490bfffc7fd10855e513e6e88be5a9a3f789" "/home/kevinweiss/WorkingDirectory/RIOT/dist/tools/edbg/bin"
Cloning into '/home/kevinweiss/WorkingDirectory/RIOT/dist/tools/edbg/bin'...
remote: Enumerating objects: 29, done.
remote: Counting objects: 100% (29/29), done.
remote: Compressing objects: 100% (24/24), done.
remote: Total 558 (delta 12), reused 14 (delta 5), pack-reused 529
Receiving objects: 100% (558/558), 232.10 KiB | 0 bytes/s, done.
Resolving deltas: 100% (354/354), done.
Checking connectivity... done.
HEAD is now at 4f5d490... Moved initial reset into target-specific code
touch /home/kevinweiss/WorkingDirectory/RIOT/dist/tools/edbg/bin/.git-downloaded
if [ 4f5d490bfffc7fd10855e513e6e88be5a9a3f789 != 4f5d490bfffc7fd10855e513e6e88be5a9a3f789 ] ; then \
	git -C /home/kevinweiss/WorkingDirectory/RIOT/dist/tools/edbg/bin clean -xdff ; \
	git -C /home/kevinweiss/WorkingDirectory/RIOT/dist/tools/edbg/bin fetch "https://github.com/ataradov/edbg" "4f5d490bfffc7fd10855e513e6e88be5a9a3f789" ; \
	git -C /home/kevinweiss/WorkingDirectory/RIOT/dist/tools/edbg/bin checkout -f 4f5d490bfffc7fd10855e513e6e88be5a9a3f789 ; \
	touch /home/kevinweiss/WorkingDirectory/RIOT/dist/tools/edbg/bin/.git-downloaded ; \
fi
env -i PATH="/home/kevinweiss/esp/esptool-esp8266-rtos-sdk-v3:/home/kevinweiss/esp/xtensa-esp8266-elf/bin:/home/kevinweiss/opt/gnu-mcu-eclipse/openocd/0.10.0-12-20190422-2015/bin/:/home/kevinweiss/bin:/home/kevinweiss/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/home/kevinweiss/opt/gcc-arm-none-eabi-7-2017-q4-major/bin:/home/kevinweiss/opt/dfu-util/src:/home/kevinweiss/tmp/pic32prog:/home/kevinweiss/esp/xtensa-esp32-elf/bin" TERM="xterm-256color" "make" -C "/home/kevinweiss/WorkingDirectory/RIOT/dist/tools/edbg/bin"
make: Entering directory '/home/kevinweiss/WorkingDirectory/RIOT/dist/tools/edbg/bin'
gcc -W -Wall -Wextra -O2 -std=gnu11 dap.c edbg.c target.c target_atmel_cm0p.c target_atmel_cm3.c target_atmel_cm4.c target_atmel_cm7.c target_atmel_cm4v2.c target_mchp_cm23.c  dbg_lin.c -ludev -o edbg
make: Leaving directory '/home/kevinweiss/WorkingDirectory/RIOT/dist/tools/edbg/bin'
mv /home/kevinweiss/WorkingDirectory/RIOT/dist/tools/edbg/bin/edbg .
[INFO] edbg binary successfully built!
/home/kevinweiss/WorkingDirectory/RIOT/dist/tools/edbg/edbg  --target atmel_cm4v2 --verbose --file /home/kevinweiss/WorkingDirectory/RIOT/tests/driver_hd44780/bin/same54-xpro/tests_driver_hd44780.bin --verify || /home/kevinweiss/WorkingDirectory/RIOT/dist/tools/edbg/edbg  --target atmel_cm4v2 --verbose --file /home/kevinweiss/WorkingDirectory/RIOT/tests/driver_hd44780/bin/same54-xpro/tests_driver_hd44780.bin --verify --program
Debugger: ATMEL EDBG CMSIS-DAP ATML2748051800004069 03.25.01B6 (S)
Clock frequency: 16.0 MHz
Target: SAM E54P20A (Rev A)
Verification...
at address 0x0 expected 0x00, read 0xf0
Error: verification failed
Debugger: ATMEL EDBG CMSIS-DAP ATML2748051800004069 03.25.01B6 (S)
Clock frequency: 16.0 MHz
Target: SAM E54P20A (Rev A)
Programming..... done.
Verification.......................... done.
/home/kevinweiss/WorkingDirectory/RIOT/dist/tools/pyterm/pyterm -p "/dev/ttyACM0" -b "115200" --noprefix --no-repeat-command-on-empty-line
Twisted not available, please install it if you want to use pyterm's JSON capabilities
Connect to serial port /dev/ttyACM0
Welcome to pyterm!
Type '/exit' to exit.
main(): This is RIOT! (Version: 2020.01-devel-443-g0c18788-tstxxx/12634)
[START]
[SUCCESS]
```

**Comment by aabadie** [comment]
Thanks @benpicco! That was it. Good catch!

**Comment by aabadie** [comment]
Another solution is to declare the device descriptor static in the test application.

**Comment by aabadie** [comment]
> Another solution is to declare the device descriptor static in the test application.

I agree this is not nice. But it is showing the potential initialization difference between structures initialized on the stack and structures defined statically.

**Comment by aabadie** [comment]
I reworked the PR with the simple fix proposed by @benpicco. AFAICT, everything works locally with this patch.

**Comment by benpicco** [comment]
I still liked the change to just check if the 4th pin is set instead of counting the pins.
The driver only works for consecutive sets of 4 or 8 pins after all.

**Comment by aabadie** [comment]
> The driver only works for consecutive sets of 4 or 8 pins after all.

Indeed. And it was saving some memory on ROM. I'll push a commit for this.

**Comment by aabadie** [comment]
Done. With this we save 16B on ROM for Arduino-Zero, not that much but better than before.

**Pull request merged**